MC Onboarding: Enabled currencies missing currency bug

tpaksu opened the pull request:
#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

**Note:** This is a follow-up PR to solve a problem in the MC onboarding code (#2786), which I didn't realize before.

The function responsible for creating concatenated strings about previously enabled currencies were editing the enabled currencies array, and it caused missing enabled currencies when submitted.

The utility is used like this: 

```jsx
<strong>
	{ ConcatenateCurrencyStrings(
		enabledCurrencyCodes,
		defaultCurrencyCode,
		availableCurrencies
	) }
</strong>
```

And `ConcatenateCurrencyStrings` had this statement inside:


```js
export const ConcatenateCurrencyStrings = (
	currencies,
	except,
	currenciesData
) => {
	if ( currencies.includes( except ) )
		currencies.splice( currencies.indexOf( except ), 1 );
```

Which altered the original array, and caused problems like missing currencies on the enabled list. This PR fixes that.

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

I don't have a concrete scenario for this but I've observed this multiple times with the previous version:

* Set your store currency something different than USD.
* Go to the onboarding screen by clicking the MC note on WCPay home screen.
* Add some currencies (including USD) and make it to the next step.
* Refresh the page and return back to step 1.
* Select another currency and save it again.
* In the previous version, you'd have USD missing from the enabled list and available back for selection again, I don't know why.
* In the current version, you shouldn't be losing any data when something like this occurs.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
-->

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

ismaeldcom commented:
Thanks for noticing that issue @tpaksu. Like I mentioned might be better to add some tests to those two functions in `utils.js`, will be easy to write the tests scenarios after having this failing case.

I've been able to reproduce the issue, and it works good after your change. 🎉 

Apart from that, I think we could simplify the code of `ConcatenateCurrencyStrings` a bit. I didn't test it, but pretty sure something like this would work:
```js
// We don't need to any early check of length, cause it will return an empty array that renders to nothing :)
return currencies
    .filter( ( code ) => code !== except ) // You already have this.
    .map( ( code ) =>
        StringRepresentationOfCurrency( currenciesData[ code ] )
    ) // We can use only one map for this part.
    .join( ', ' )
    .replace( /,([^,]*)$/, ` ${ __( 'and', 'woocommerce-payments' ) }$1` ); // This will replace only the last occurrence of ", "
```

tpaksu commented:
Thanks @ismaeldcom for the review, I changed your code a bit because your code continued to add the Oxford comma to the strings which have two elements, AFAIK they shouldn't have that. Otherwise, yes, I could simplify the function, so did I. Thanks for recommending me to add tests too, it helped me while I refactored the code too!

ismaeldcom commented on client/multi-currency-setup/tasks/add-currencies-task/test/utils.test.js:
```diff
@@ -0,0 +1,172 @@
+		symbol: '¥',
+	},
+	INR: {
+		code: 'INR',
+		rate: 1,
+		name: 'Indian rupee',
+		id: 'inr',
+		is_default: false,
+		flag: '🇮🇳',
+		symbol: '₹',
+		is_zero_decimal: false,
+		last_updated: 1630070442,
+	},
+	DKK: {
+		code: 'DKK',
+		rate: '6.144615',
+		name: 'Danish krone',
+		id: 'dkk',
+		is_default: false,
+		flag: '🇩🇰',
+		symbol: 'DKK',
+	},
+	BIF: {
+		code: 'BIF',
+		rate: '1974',
+		name: 'Burundian franc',
+		id: 'bif',
+		is_default: false,
+		flag: '🇧🇮',
+		symbol: 'Fr',
+	},
+	CLP: {
+		code: 'CLP',
+		rate: '706.8',
+		name: 'Chilean peso',
+		id: 'clp',
+		is_default: false,
+		flag: '🇨🇱',
+		symbol: '$',
+	},
```
Don't want to be too picky 😅 But we can get rid of the unused currencies properties in the test:
```suggestion
	USD: {
		code: 'USD',
		name: 'United States (US) dollar',
		symbol: '$',
	},
	CAD: {
		code: 'CAD',
		name: 'Canadian dollar',
		symbol: '$',
	},
	GBP: {
		code: 'GBP',
		name: 'Pound sterling',
		symbol: '£',
	},
	EUR: {
		code: 'EUR',
		name: 'Euro',
		symbol: '€',
	},
	AED: {
		code: 'AED',
		name: 'United Arab Emirates dirham',
		symbol: 'د.إ',
	},
	CDF: {
		code: 'CDF',
		name: 'Congolese franc',
		symbol: 'Fr',
	},
	AUD: {
		code: 'AUD',
		name: 'Australian dollar',
		symbol: '$',
	},
	JPY: {
		code: 'JPY',
		name: 'Japanese yen',
		symbol: '¥',
	},
	INR: {
		code: 'INR',
		name: 'Indian rupee',
		symbol: '₹',
	},
	DKK: {
		code: 'DKK',
		name: 'Danish krone',
		symbol: 'DKK',
	},
	BIF: {
		code: 'BIF',
		name: 'Burundian franc',
		symbol: 'Fr',
	},
	CLP: {
		code: 'CLP',
		name: 'Chilean peso',
		symbol: '$',
	},
```

tpaksu replied:
Sorry, I merged it before seeing this comment 😮 I guess they'll stay 😬

ismaeldcom reviewed (approved):
LGTM! Thanks for adding tests and update the function code. Let's 🚢  it!
